Navigating secrets consumption in Kubernetes pods blog post (#17)
handsonarchitects/blog: content/blog/2024/navigating-secrets-consumption-in-kubernetes-pods/index.md

## Secret as Environment Variables (secretKeyRef)

The final method for consuming secrets in Kubernetes is by injecting them as environment variables using the `secretKeyRef` field in the pod specification. This approach is suitable when your application requires access to specific keys from the secret as environment variables.

1. Create a pod with the secret injected as environment variables using `secretKeyRef`.
```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: pod-with-secret-env-secret-key-ref
  name: pod-with-secret-env-secret-key-ref
spec:
    containers:
    - image: nginx
      name: pod-with-secret-env-secret-key-ref
      env:
      - name: DB_PASSWORD
        valueFrom:
          secretKeyRef:
            name: db-secret
            key: DB_PASSWORD
```
2. Check the status of the pod.
```bash
kubectl get pods | grep pod-with-secret-env-secret-key-ref
```
The pod will not be created until the secret is present.

3. Create a secret with a database credentials.
```bash
kubectl create secret generic db-secret --from-literal=DB_ADMIN=myadmin --from-literal=DB_PASSWORD=mypass
```
4. Check the status of the pod.
```bash

> kubectl get pods | grep pod-with-secret-env-secret-key-ref

Reviewer: ```suggestion
kubectl get pod pod-with-secret-env-secret-key-ref
```

> kubectl exec -it pod-with-secret-env-secret-key-ref  -- env | grep DB_

Reviewer: ```suggestion
kubectl exec -it pod-with-secret-env-secret-key-ref  -- env | grep APP_PREFIX_
```

```
The expected result is:
```bash
pod-with-secret-env-secret-key-ref   1/1     Running...
...

> DB_PASSWORD=mypass

Reviewer: ```suggestion
APP_PREFIX_DB_PASS=mypass
```

```
5. Delete the secret and observe the behavior of the pod.
```bash
kubectl delete secret db-secret

> kubectl get pods | grep pod-with-secret-env-secret-key-ref

Reviewer: ```suggestion
kubectl get pod pod-with-secret-env-secret-key-ref
```

> kubectl exec -it pod-with-secret-env-secret-key-ref  -- env | grep DB_

Reviewer: ```suggestion
kubectl exec -it pod-with-secret-env-secret-key-ref  -- env | grep APP_PREFIX
```

```
The expected result is:
```bash
pod-with-secret-env-secret-key-ref   1/1     Running...
...

> DB_PASSWORD=mypass

Reviewer: ```suggestion
APP_PREFIX_DB_PASS=mypass
```

```
6. Recreate the secret with a new database password.
```bash
kubectl create secret generic db-secret --from-literal=DB_ADMIN=myadmin --from-literal=DB_PASSWORD=mynewpass
```
7. View the environment variables in the pod.
```bash

> kubectl exec -it pod-with-secret-env-secret-key-ref  -- env | grep DB_

Reviewer: ```suggestion
kubectl exec -it pod-with-secret-env-secret-key-ref  -- env | grep APP_PREFIX
```

```
The result should be:
```bash

> DB_PASSWORD=mypass

Reviewer: ```suggestion
APP_PREFIX_DB_PASS=mypass
```

```
As observed in this scenario, the pod is not affected if the secret is deleted. Container environment variables are not updated when the secret is recreated.

When to Use Secret as Environment Variables (secretKeyRef):
- when your application requires access to specific keys from the secret as environment variables (more control over the keys, more secure)
- when the pod should not be created if the specific key is not present in the secret during the pod's creation

> The difference between the `envFrom` and `secretKeyRef` methods is that the `envFrom` method loads all the key-value pairs from the secret as environment variables, while the `secretKeyRef` method allows you to specify the keys from the secret that should be injected as environment variables. **If the key is not present in the secret, the pod will not be created (during pod creation time).**

Reviewer: ```suggestion
The difference between the `envFrom` and `secretKeyRef` methods is that the `envFrom` method loads all the key-value pairs from the secret as environment variables, while the `secretKeyRef` method allows you to specify the keys from the secret that should be injected as environment variables. With the default settings, **if the key is not present in the secret, the pod will not be created (during pod creation time).**
```

## Summary

The consumption of secrets in Kubernetes pods and containers in all scenarios above follows similar behavior patterns.

> - the pod is not created until the secret is present

Reviewer: ```suggestion
- by default, the pod is not created until the secret is present
```
